### PR TITLE
Cancel active fetching on a new route request

### LINKED ISSUE #2825: Handle multiple routes requests
in 1Tap we use coroutines to fetch routes with `mapboxNavigation.requestRoutes(options, callback)`. The coroutine is suspended while fetching is in progress. We resume it in the callback's `onRoutesReady/onRoutesRequestCanceled/onRoutesRequestFailure`.
In some cases we get
`java.lang.IllegalStateException: Already resumed at kotlin.coroutines.SafeContinuation.resumeWith(SafeContinuationJvm.kt:45)`
it means that the same callback was called `twice` from the SDK side and tried to resume the same coroutine twice.
After some investigation I figured out that it happens only if we try to fetch a route while another fetching is in progress.
Looks like the next happens:
1. start `fetching_1` and pass `callback_1`
2. start `fetching_2` and pass `callback_2`
3. SDK overrides `callback_1` with `callback_2`
4. SDK sends success to `callback_2` as a result of `fetching_2` -> OK! resume coroutine
5. SDK sends canceled to `callback_2` as a result of `fetching_1` -> error, `callback_1` have to be used -> get `IllegalStateException`
there are some logs:
2020-04-22 13:47:22.075 : requestRoutesWith lambda$1@86855c to Point{type=Point, bbox=null, coordinates=[49.57386815303724, 65.35148780980239]}
2020-04-22 13:47:23.054 : requestRoutesWith lambda$1@c7d4b53 to Point{type=Point, bbox=null, coordinates=[50.49439852690722, 62.012377288968906]}
2020-04-22 13:47:23.127 : onRoutesReady lambda$1@c7d4b53
2020-04-22 13:47:23.141 : onRoutesRequestCanceled lambda$1@c7d4b53 to Point{type=Point, bbox=null, coordinates=[50.49439852690722, 62.012377288968906]}
as you can see,
`lambda$1@86855c` was not called at all
`lambda$1@c7d4b53` called twice
possible solutions on consumer side:
as a quick fix, we can call something like `mapboxNavigation.cancelRouteFetching()` before each requestRoutes call, but we don't have such public method
also we can just disable a new call until previous is finished (disable view interactions)
on SDK side:
handle current callback and a new one?

### PR DESCRIPTION
**Pull request by korshaknn**
## Description

Check description [here](https://github.com/mapbox/mapbox-navigation-android/issues/2825)
closes #2825 

- [x] I have added any issue links
- [x] I have added all related labels (`bug`, `feature`, `new API(s)`, `SEMVER`, etc.)
- [x] I have added the appropriate milestone and project boards

## Testing

Please describe the manual tests that you ran to verify your changes

- [ ] I have tested locally (including `SNAPSHOT` upstream dependencies if needed) through testapp/demo app and run all activities to avoid regressions
- [ ] I have tested via a test drive, or a simulation/mock location app
- [ ] I have added tests that prove my fix is effective or that my feature works
- [ ] New and existing unit tests pass locally with my changes

## Checklist

- [ ] My code follows the style guidelines of this project
- [ ] I have performed a self-review of my own code
- [ ] I have updated the `CHANGELOG` including this PR

### CODE REVIEW

File: libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt [hunk 1]
```diff
@@ -75,6 +75,7 @@ class MapboxHybridRouter(
         private var reserveRouterCalled = false
         private lateinit var options: RouteOptions
         private lateinit var callback: Router.Callback
+        private var fetchingInProgress = false
 
         override fun onResponse(routes: List<DirectionsRoute>) {
             callback.onResponse(routes)
```
Comment (Guardiola31337) [inline libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt:81]: I believe we should also make `fetchingInProgress` `false` when receiving a response, right?

Comment (korshaknn) [reply libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt:82]: fixed

File: libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt [hunk 2]
```diff
@@ -101,22 +107,41 @@ class MapboxHybridRouter(
...
 
         override fun getRouteRefresh(route: DirectionsRoute, legIndex: Int, callback: RouteRefreshCallback) {
             mainRouter.getRouteRefresh(route, legIndex, callback)
         }
+
+        private fun checkPendingRequest() {
+            if (pendingRequest != null) {
+                getRoute(pendingRequest!!.first, pendingRequest!!.second)
+                pendingRequest = null
+            }
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt]: let's use param names
```
pendingRequest?.let {
      getRoute(routeOptions = it.first, clientCallback = it.second)
 }
```

File: libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt [hunk 3]
```diff
@@ -101,22 +107,41 @@ class MapboxHybridRouter(
...
+                } else {
+                    mainRouter.cancel()
+                }
+            } else {
+                fetchingInProgress = true
+                reserveRouterCalled = false
+                options = routeOptions
+                callback = clientCallback
+                mainRouter.getRoute(routeOptions, this)
+            }
```
[the remark this answers is not in the document]
Comment (korshaknn) [reply libdirections-hybrid/src/main/java/com/mapbox/navigation/route/hybrid/MapboxHybridRouter.kt:132]: done

**Comment by LukasPaczos** [comment]
Nice work @korshaknn! Could you follow up with a unit test?

**Pull request merged**